hack/generator/pkg/astmodel/package_definition.go

```diff
@@ -84,131 +77,37 @@ func emitFiles(filesToGenerate map[string][]TypeDefinition, outputDir string) er
 	return nil
 }
 
-func anyReferences(defs []TypeDefinition, defName TypeName) bool {
-	for _, def := range defs {
-		if def.References().Contains(defName) {
-			return true
-		}
-	}
-
-	return false
-}
-
-type partitionedDefinitions struct {
-	resources        []TypeDefinition
-	otherDefinitions []TypeDefinition
-}
-
-func partitionDefinitions(definitions []TypeDefinition) partitionedDefinitions {
-	var result partitionedDefinitions
-
-	for _, def := range definitions {
-		if _, ok := def.Type().(*ResourceType); ok {
-			result.resources = append(result.resources, def)
-		} else {
-			result.otherDefinitions = append(result.otherDefinitions, def)
-		}
-	}
-
-	return result
-}
-
-func allocateTypesToFiles(typesToAllocate []TypeDefinition, filesToGenerate map[string][]TypeDefinition) {
-	// Keep track of how many types we've checked since we last allocated one
-	// This lets us detect if/when we stall during allocation
-	skippedTypes := 0
-
-	// Limit to how many types we skip before taking remedial action
-	// len(typesToAllocate) changes during execution, so we cache it
-	skipLimit := len(typesToAllocate)
-
-	for len(typesToAllocate) > 0 {
-		// dequeue!
-		typeToAllocate := typesToAllocate[0]
-		typesToAllocate = typesToAllocate[1:]
-
-		filesReferencingType := findFilesReferencingType(typeToAllocate, filesToGenerate)
-		hasPendingReferences := anyReferences(typesToAllocate, typeToAllocate.Name())
+func allocateTypesToFiles(types Types) map[string][]TypeDefinition {
+	graph := NewReferenceGraphWithResourcesAsRoots(types)
 
-		allocatedFileName := allocateFileName(filesReferencingType, typeToAllocate, hasPendingReferences)
-		if allocatedFileName == "" && skippedTypes > skipLimit {
-			// We've stalled during allocation of types, implying we have a cycle of mutually referencing types
-			allocatedFileName = allocateFilenameWhenStalled(typesToAllocate, filesToGenerate, len(filesReferencingType), typeToAllocate)
+	// rootsFor maps a type name to the roots that can reach it
+	rootsFor := make(map[TypeName][]TypeName)
+	for root := range graph.roots {
+		collected := make(TypeNameSet)
+		graph.CollectTypes(root, collected)
```

Comment:
That method is public here? So maybe it does need to be public?

Reply:
it’s in the same package